File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/CommitCollector.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit;
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile;
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collector;

class CommitCollector {

    private final MetricsFactory metricsFactory;

    private CommitCollector(MetricsFactory metricsFactory) {
        this.metricsFactory = metricsFactory;
    }

    static Collector<Commit, ?, List<VersionControlledFile>> create(MetricsFactory metricsFactory) {
        CommitCollector collector = new CommitCollector(metricsFactory);
        return Collector.of(ArrayList::new, collector::collectCommit, collector::combineForParallelExecution);
    }

    private void collectCommit(List<VersionControlledFile> versionControlledFiles, Commit commit) {
        if (commit.isEmpty()) {
            return;
        }
        addYetUnknownFilesToVersionControlledFiles(versionControlledFiles, commit.getFilenames());
        addCommitMetadataToMatchingVersionControlledFiles(commit, versionControlledFiles);
    }

    private void addYetUnknownFilesToVersionControlledFiles(List<VersionControlledFile> versionControlledFiles, List<String> filenamesOfCommit) {
        filenamesOfCommit.stream()
                .filter(filename -> !versionControlledFilesContainsFile(versionControlledFiles, filename))
                .forEach(unknownFilename -> addYetUnknownFile(versionControlledFiles, unknownFilename));
    }

    private boolean versionControlledFilesContainsFile(List<VersionControlledFile> versionControlledFiles, String filename) {
        return findVersionControlledFileByFilename(versionControlledFiles, filename).isPresent();
    }

    private Optional<VersionControlledFile> findVersionControlledFileByFilename(List<VersionControlledFile> versionControlledFiles, String filename) {
        return versionControlledFiles.stream()
                .filter(commit -> commit.getFilename().equals(filename))
                .findFirst();
    }

    private boolean addYetUnknownFile(List<VersionControlledFile> versionControlledFiles, String filenameOfYetUnversionedFile) {
        VersionControlledFile missingVersionControlledFile = new VersionControlledFile(filenameOfYetUnversionedFile, metricsFactory);
        return versionControlledFiles.add(missingVersionControlledFile);
    }

    private void addCommitMetadataToMatchingVersionControlledFiles(Commit commit, List<VersionControlledFile> versionControlledFiles) {
        commit.getFilenames().stream()
                .map(file -> findVersionControlledFileByFilename(versionControlledFiles, file))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .forEach(vcFile -> vcFile.registerCommit(commit));

    }

    private List<VersionControlledFile> combineForParallelExecution(List<VersionControlledFile> firstCommits, List<VersionControlledFile> secondCommits) {
        throw new UnsupportedOperationException("parallel collection of commits not supported");
    }

}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/CommitCollector.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory
import java.util.stream.Collector

internal class CommitCollector private constructor(private val metricsFactory: MetricsFactory) {

    private fun collectCommit(versionControlledFiles: MutableList<VersionControlledFile>, commit: Commit) {
        if (commit.isEmpty) {
            return
        }
        addYetUnknownFilesToVersionControlledFiles(versionControlledFiles, commit.filenames)
        addCommitMetadataToMatchingVersionControlledFiles(commit, versionControlledFiles)
    }

    private fun addYetUnknownFilesToVersionControlledFiles(
        versionControlledFiles: MutableList<VersionControlledFile>,
        filenamesOfCommit: List<String>
    ) {
        filenamesOfCommit
            .filter { !versionControlledFilesContainsFile(versionControlledFiles, it) }
            .forEach { addYetUnknownFile(versionControlledFiles, it) }
    }

    private fun versionControlledFilesContainsFile(
        versionControlledFiles: List<VersionControlledFile>,
        filename: String
    ): Boolean {
        return findVersionControlledFileByFilename(versionControlledFiles, filename) != null
    }

    private fun findVersionControlledFileByFilename(
        versionControlledFiles: List<VersionControlledFile>,
        filename: String
    ): VersionControlledFile? {
        return versionControlledFiles.firstOrNull { it.filename == filename }
    }

    private fun addYetUnknownFile(
        versionControlledFiles: MutableList<VersionControlledFile>,
        filenameOfYetUnversionedFile: String
    ): Boolean {
        val missingVersionControlledFile = VersionControlledFile(filenameOfYetUnversionedFile, metricsFactory)
        return versionControlledFiles.add(missingVersionControlledFile)
    }

    private fun addCommitMetadataToMatchingVersionControlledFiles(
        commit: Commit,
        versionControlledFiles: List<VersionControlledFile>
    ) {
        commit.filenames.mapNotNull { findVersionControlledFileByFilename(versionControlledFiles, it) }
            .forEach { it.registerCommit(commit) }
    }

    companion object {

        fun create(metricsFactory: MetricsFactory): Collector<Commit, *, MutableList<VersionControlledFile>> {
            val collector = CommitCollector(metricsFactory)
            return Collector.of(
                { ArrayList() },
                { versionControlledFiles,
                    commit ->
                    collector.collectCommit(versionControlledFiles, commit)
                },
                { _, _ ->
                    throw UnsupportedOperationException("parallel collection of commits not supported")
                }
            )
        }
    }
}
